Storage/Clients.swift

```diff
@@ -49,6 +51,59 @@ public struct RemoteClient: Equatable {
         self.application = nil
         self.version = nil
         self.protocols = nil
+        self.commands = nil
+    }
+
+    public init(client: RemoteClient, withCommands commands: [SyncCommand]) {
+        self.guid = client.guid
+        self.name = client.name
+        self.modified = client.modified
+        self.type = client.type
+        self.formfactor = client.formfactor
+        self.os = client.os
+
+        self.device = client.device
+        self.appPackage = client.appPackage
+        self.application = client.application
+        self.version = client.version
+        self.protocols = client.protocols
+        self.commands = commands
+    }
+
+    func withCommands(commands: [SyncCommand]) -> RemoteClient {
+        return RemoteClient(client: self, withCommands: commands)
+    }
+
+    public func toJSON() -> JSON {
+        var jsonCommands: [JSON]
+        if let cmds = self.commands {
+            jsonCommands = cmds.map { command in
+                JSON(command.value)
+            }
+        } else {
+            jsonCommands = [JSON]()
+        }
+        let jsonFormFactor = formfactor ?? ""
+        let jsonDevice = device ?? ""
+        let jsonAppPackage = appPackage ?? ""
+        let jsonApplication = application ?? ""
+        let jsonVersion = version ?? ""
+        let jsonProtocols = protocols ?? [] 
+
+        let json = JSON([
+            "id": guid ?? "",
+            "version": jsonVersion,    // TODO
+            "protocols": jsonProtocols,
+            "name": name,
+            "os": os ?? "iOS",
```

Comment:
This is wrong for the reason I mentioned last week.

Client A uploads a client record that's missing `os`. We download it and stuff it into the DB. It ends up as a `RemoteClient` with `os = null`.

We now want to send a command to that device. We build some JSON… and here we default it to `"iOS"` + `"mobile"`! A few lines above we also trample the protocol, versions, etc.

Until that client uploads a new record, other clients will see it as an iPhone. Worse, any clients that try to use the version lists to decide on behavior will draw the wrong conclusions.

These really are optional fields, so they should be omitted unless we downloaded them in the first place.

There's another bug hiding here, too, which you should be able to write a test for. What happens if we're sending a second tab to a client that already has one queued? If I'm reading this code correctly, we'll overwrite any existing commands, including the existing `displayURI` command.

That leaves us with two options:

• Make sure that our database stores everything we need to re-upload a client record, including its existing commands array.
• Do what other clients do: work with client records at the JSON level, and use our DB only for client picking and display. When we want to send a command, do a two-step download-change-upload, mutating only the command array in the downloaded record.

I think the latter makes most sense.